File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/example2/examples/MecDrive.java
package org.firstinspires.ftc.teamcode.example2.examples;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorSimple;

import org.firstinspires.ftc.teamcode.example2.backEnd.HardwareClass;

/**
 * Created by David Lukens on 7/23/2018.
 */

public class MecDrive {
	private final DcMotor lf;
	private final DcMotor lb;
	private final DcMotor rf;
	private final DcMotor rb;

	private final DcMotor.RunMode runMode;
	private final HardwareClass   robot;

	private void setRunMode(DcMotor.RunMode runMode) {
		lf.setMode(runMode);
		lb.setMode(runMode);
		rf.setMode(runMode);
		rb.setMode(runMode);
	}

	public MecDrive(HardwareClass robot, boolean reverseLF, boolean reverseLB, boolean reverseRF, boolean reverseRB, DcMotor.RunMode runMode) {
		lf = robot.hMap.dcMotor.get("lf");
		lb = robot.hMap.dcMotor.get("lb");
		rf = robot.hMap.dcMotor.get("rf");
		rb = robot.hMap.dcMotor.get("rb");

		lf.setDirection(reverseLF ? DcMotorSimple.Direction.REVERSE : DcMotorSimple.Direction.FORWARD);
		lb.setDirection(reverseRB ? DcMotorSimple.Direction.REVERSE : DcMotorSimple.Direction.FORWARD);
		rf.setDirection(reverseLF ? DcMotorSimple.Direction.REVERSE : DcMotorSimple.Direction.FORWARD);
		rb.setDirection(reverseRB ? DcMotorSimple.Direction.REVERSE : DcMotorSimple.Direction.FORWARD);

		this.runMode = runMode;
		this.robot = robot;

		setRunMode(runMode);
	}

	public void powerDrive(double forward, double turnClockwise, double strafeRight) {
		double lf = forward + strafeRight + turnClockwise;
		double lb = forward - strafeRight + turnClockwise;
		double rf = forward - strafeRight - turnClockwise;
		double rb = forward + strafeRight - turnClockwise;

		double max = Math.max(Math.abs(lf), Math.abs(lb));
		max = Math.max(max, Math.abs(rf));
		max = Math.max(max, Math.abs(rb));

		if(max > 1.0) {
			lf /= max;
			lb /= max;
			rf /= max;
			rb /= max;
		}

		this.lf.setPower(lf);
		this.lb.setPower(lb);
		this.rf.setPower(rf);
		this.rb.setPower(rb);
	}

	public void resetEncoders() {
		setRunMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
		robot.opMode.idle();
		setRunMode(runMode);
	}
}

File: rechargedkotlinlibrary/src/main/java/com/david/rechargedkotlinlibrary/internal/hardware/driveTerrain/MecDrive.kt
package com.david.rechargedkotlinlibrary.internal.hardware.driveTerrain

import com.acmerobotics.roadrunner.drive.MecanumDrive
import com.david.rechargedkotlinlibrary.internal.hardware.devices.OptimumDcMotorEx
import com.david.rechargedkotlinlibrary.internal.hardware.management.RobotTemplate
import com.david.rechargedkotlinlibrary.internal.hardware.management.SameThreadSubsystem
import com.david.rechargedkotlinlibrary.internal.util.MathUtil
import com.qualcomm.robotcore.hardware.DcMotor
import java.util.*
import kotlin.math.abs

/**
 * Created by David Lukens on 8/2/2018.
 */
open class MecDrive(private val lf: OptimumDcMotorEx,
                    private val lb: OptimumDcMotorEx,
                    private val rf: OptimumDcMotorEx,
                    private val rb: OptimumDcMotorEx,
                    private val RUN_MODE: DcMotor.RunMode = DcMotor.RunMode.RUN_USING_ENCODER,
                    zeroPowerBehavior: DcMotor.ZeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE,
                    private val ENCODER_SCALER: Double = 1.0,
                    private val RADIUS: Double = 2.0,
                    TRACK_WIDTH: Double,
                    WHEEL_BASE: Double)
    : MecanumDrive(TRACK_WIDTH, WHEEL_BASE) {
    init {
        lf.mode = RUN_MODE
        lb.mode = RUN_MODE
        rf.mode = RUN_MODE
        rb.mode = RUN_MODE
        lf.zeroPowerBehavior = zeroPowerBehavior
        lb.zeroPowerBehavior = zeroPowerBehavior
        rf.zeroPowerBehavior = zeroPowerBehavior
        rb.zeroPowerBehavior = zeroPowerBehavior
    }

    fun powerTranslation(forward: Double, strafeRight: Double, turnClockwise: Double) = setMotorPowers(forward + strafeRight + turnClockwise, forward - strafeRight + turnClockwise, forward - strafeRight - turnClockwise, forward + strafeRight - turnClockwise)

    override fun setMotorPowers(frontLeft: Double, rearLeft: Double, rearRight: Double, frontRight: Double) {
        val max = Collections.max(listOf(abs(frontLeft), abs(rearLeft), abs(frontRight), abs(rearRight), 1.0))
        lf.power = frontLeft / max
        lb.power = rearLeft / max
        rf.power = frontRight / max
        rb.power = rearRight / max
    }

    fun radiansToInches(radians: Double) = MathUtil.radiansToInches(radians * ENCODER_SCALER, RADIUS)

    override fun getWheelPositions() = ListOf(
            radiansToInches(lfRawRadians()),
            radiansToInches(lbRawRadians()),
            radiansToInches(rfRawRadians()),
            radiansToInches(rbRawRadians()))

    fun resetEncoders() {
        resetLFEncoder()
        resetLBEncoder()
        resetRFEncoder()
        resetRBEncoder()
    }

    fun resetLFEncoder() = lf.resetEncoder()
    fun resetLBEncoder() = lb.resetEncoder()
    fun resetRFEncoder() = rf.resetEncoder()
    fun resetRBEncoder() = rb.resetEncoder()
    fun lfTicks() = lf.currentPosition
    fun lbTicks() = lb.currentPosition
    fun rfTicks() = rf.currentPosition
    fun rbTicks() = rb.currentPosition
    fun lfRawTicks() = lf.getRawPosition()
    fun lbRawTicks() = lb.getRawPosition()
    fun rfRawTicks() = rf.getRawPosition()
    fun rbRawTicks() = rb.getRawPosition()
    fun lfRadians() = lf.getRadians()
    fun lbRadians() = lb.getRadians()
    fun rfRadians() = rf.getRadians()
    fun rbRadians() = rb.getRadians()
    fun lfRawRadians() = lf.getRawRadians()
    fun lbRawRadians() = lb.getRawRadians()
    fun rfRawRadians() = rf.getRawRadians()
    fun rbRawRadians() = rb.getRawRadians()
}